src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/AppointmentService.kt

```diff
@@ -14,54 +16,67 @@ import javax.transaction.Transactional
 @Transactional
 class AppointmentService(
   val appointmentRepository: AppointmentRepository,
+  val communityAPIBookingService: CommunityAPIBookingService,
   val authUserRepository: AuthUserRepository,
 ) {
   fun createOrUpdateAppointment(
+    referral: Referral,
     appointment: Appointment?,
     durationInMinutes: Int,
     appointmentTime: OffsetDateTime,
+    appointmentType: AppointmentType,
     createdByUser: AuthUser
   ): Appointment {
-    val noAppointment = appointment == null
-    if (noAppointment) {
-      return createAppointment(durationInMinutes, appointmentTime, createdByUser)
+
+    val initialAppointmentRequired = appointment == null
+    if (initialAppointmentRequired) {
+      val deliusAppointmentId = communityAPIBookingService.book(referral, null, appointmentTime, durationInMinutes, appointmentType)
+      val createdAppointment = createAppointment(durationInMinutes, appointmentTime, deliusAppointmentId, createdByUser)
+      return appointmentRepository.save(createdAppointment)
     }
 
-    val appointmentWithoutAttendanceRecorded = appointment!!.attended == null
-    if (appointmentWithoutAttendanceRecorded) {
-      return updateAppointment(appointment, durationInMinutes, appointmentTime)
+    val updateCurrentAppointmentRequired = appointment!!.attended == null
+    if (updateCurrentAppointmentRequired) {
+      val deliusAppointmentId = communityAPIBookingService.book(referral, appointment, appointmentTime, durationInMinutes, appointmentType)
+      val updatedAppointment = updateAppointment(appointment, durationInMinutes, appointmentTime, deliusAppointmentId)
+      return appointmentRepository.save(updatedAppointment)
     }
 
-    val appointmentWithNonAttendance = appointment.attended == Attended.NO
-    if (appointmentWithNonAttendance) {
-      return createAppointment(durationInMinutes, appointmentTime, createdByUser)
+    val additionalAppointmentRequired = appointment.attended == Attended.NO
+    if (additionalAppointmentRequired) {
+      val deliusAppointmentId = communityAPIBookingService.book(referral, null, appointmentTime, durationInMinutes, appointmentType)
+      val additionalAppointment = createAppointment(durationInMinutes, appointmentTime, deliusAppointmentId, createdByUser)
+      return appointmentRepository.save(additionalAppointment)
     }
+
     throw IllegalStateException("Is it not possible to update an appointment that has already been attended")
```

Comment:
Could we make this check at the very beginning as a pre-condition? It feels important to be able to clearly express, however, currently, someone has to follow through all branches and build the negation of all possible conditions

Reply:
It can only arise if we are 1) not making a fresh booking 2) not doing a reschedule for an unattended appointment and 3) not creating an additional appointment for a missed appointment, i.e. should never arise. 
Expressing 1-3 as a pre condition (rather than a fall through "all other situations") I think would be overly complicated to reason